src/shims/x86/aesni.rs

```diff
@@ -0,0 +1,384 @@
+        0x72, 0xF8, 0xF6, 0x64, 0x86, 0x68, 0x98, 0x16, 0xD4, 0xA4, 0x5C, 0xCC, 0x5D, 0x65, 0xB6, 0x92,
+        0x6C, 0x70, 0x48, 0x50, 0xFD, 0xED, 0xB9, 0xDA, 0x5E, 0x15, 0x46, 0x57, 0xA7, 0x8D, 0x9D, 0x84,
+        0x90, 0xD8, 0xAB, 0x00, 0x8C, 0xBC, 0xD3, 0x0A, 0xF7, 0xE4, 0x58, 0x05, 0xB8, 0xB3, 0x45, 0x06,
+        0xD0, 0x2C, 0x1E, 0x8F, 0xCA, 0x3F, 0x0F, 0x02, 0xC1, 0xAF, 0xBD, 0x03, 0x01, 0x13, 0x8A, 0x6B,
+        0x3A, 0x91, 0x11, 0x41, 0x4F, 0x67, 0xDC, 0xEA, 0x97, 0xF2, 0xCF, 0xCE, 0xF0, 0xB4, 0xE6, 0x73,
+        0x96, 0xAC, 0x74, 0x22, 0xE7, 0xAD, 0x35, 0x85, 0xE2, 0xF9, 0x37, 0xE8, 0x1C, 0x75, 0xDF, 0x6E,
+        0x47, 0xF1, 0x1A, 0x71, 0x1D, 0x29, 0xC5, 0x89, 0x6F, 0xB7, 0x62, 0x0E, 0xAA, 0x18, 0xBE, 0x1B,
+        0xFC, 0x56, 0x3E, 0x4B, 0xC6, 0xD2, 0x79, 0x20, 0x9A, 0xDB, 0xC0, 0xFE, 0x78, 0xCD, 0x5A, 0xF4,
+        0x1F, 0xDD, 0xA8, 0x33, 0x88, 0x07, 0xC7, 0x31, 0xB1, 0x12, 0x10, 0x59, 0x27, 0x80, 0xEC, 0x5F,
+        0x60, 0x51, 0x7F, 0xA9, 0x19, 0xB5, 0x4A, 0x0D, 0x2D, 0xE5, 0x7A, 0x9F, 0x93, 0xC9, 0x9C, 0xEF,
+        0xA0, 0xE0, 0x3B, 0x4D, 0xAE, 0x2A, 0xF5, 0xB0, 0xC8, 0xEB, 0xBB, 0x3C, 0x83, 0x53, 0x99, 0x61,
+        0x17, 0x2B, 0x04, 0x7E, 0xBA, 0x77, 0xD6, 0x26, 0xE1, 0x69, 0x14, 0x63, 0x55, 0x21, 0x0C, 0x7D,
+    ];
+    s.map(|byte| INV_SBOX[usize::from(byte)])
+}
+
+/// MixColumns - multiplies `MATRIX` by `s` in GF(2^8)
+fn mix_columns(s: [u8; 16]) -> [u8; 16] {
+    #[rustfmt::skip]
+    const MATRIX: [u8; 16] = [
+        0x02, 0x01, 0x01, 0x03,
+        0x03, 0x02, 0x01, 0x01,
+        0x01, 0x03, 0x02, 0x01,
+        0x01, 0x01, 0x03, 0x02,
+    ];
+    mul_matrix_gf2p8(MATRIX, s)
+}
+
+/// InvMixColumns - multiplies `INV_MATRIX` by `s` in GF(2^8)
+fn inv_mix_columns(s: [u8; 16]) -> [u8; 16] {
+    #[rustfmt::skip]
+    const INV_MATRIX: [u8; 16] = [
+        0x0E, 0x09, 0x0D, 0x0B,
+        0x0B, 0x0E, 0x09, 0x0D,
+        0x0D, 0x0B, 0x0E, 0x09,
+        0x09, 0x0D, 0x0B, 0x0E,
+    ];
+    mul_matrix_gf2p8(INV_MATRIX, s)
+}
+
+/// Matrix multiplication where each element is a GF(2^8) polynomial
+fn mul_matrix_gf2p8(a: [u8; 16], b: [u8; 16]) -> [u8; 16] {
+    fn index(row: usize, col: usize) -> usize {
+        // Calculate index of element at (row,col)
+        // Remember that elements are arranged in column-major order.
+        col.checked_mul(4).unwrap().checked_add(row).unwrap()
+    }
+
+    let mut res = [0; 16];
+    for row in 0..4 {
+        for col in 0..4 {
+            let res = &mut res[index(row, col)];
+            *res ^= mul_gf2p8(a[index(row, 0)], b[index(0, col)]);
+            *res ^= mul_gf2p8(a[index(row, 1)], b[index(1, col)]);
+            *res ^= mul_gf2p8(a[index(row, 2)], b[index(2, col)]);
+            *res ^= mul_gf2p8(a[index(row, 3)], b[index(3, col)]);
+        }
+    }
+    res
+}
+
+/// SubWord - S-box transformation on 4 bytes
+fn sub_word(w: u32) -> u32 {
+    u32::from_ne_bytes(w.to_ne_bytes().map(|byte| SBOX[usize::from(byte)]))
```

Comment:
I would strongly caution against any implementation of AES which uses lookup tables like this. They are notorious for cache timing and other microarchitectural sidechannels:

https://eprint.iacr.org/2018/1002.pdf

Modern implementations of AES, such as the one in the `aes` crate, use techniques like bitslicing/fixslicing to avoid such sidechannels.

Reply:
Nobody should ever run Miri on secret data anyway. So if this ever becomes relevant something already went seriously wrong.

But indeed that's one of the reasons why I don't want to maintain an AES implementation.^^